src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1045 +207,1027 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
-                OnAfterAccess(args);
+            OnBeforeAccess(args);
+            // filtered by client id.
+            ICollection<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensForClient(requestParams.RequestContext);
+            OnAfterAccess(args);
 
-                // this is OBO flow. match the cache entry with assertion hash,
-                // Authority, ScopeSet and client Id.
-                if (requestParams.UserAssertion != null)
+            // this is OBO flow. match the cache entry with assertion hash,
+            // Authority, ScopeSet and client Id.
+            if (requestParams.UserAssertion != null)
+            {
+                requestParams.RequestContext.Logger.Info("Filtering by user assertion...");
+                tokenCacheItems =
+                    tokenCacheItems.Where(
+                            item =>
+                                !string.IsNullOrEmpty(item.UserAssertionHash) &&
+                                item.UserAssertionHash.Equals(requestParams.UserAssertion.AssertionHash, StringComparison.OrdinalIgnoreCase))
+                        .ToList();
+            }
+            else
+            {
+                if (!requestParams.IsClientCredentialRequest)
                 {
-                    requestParams.RequestContext.Logger.Info("Filtering by user assertion...");
+                    requestParams.RequestContext.Logger.Info("Filtering by user identifier...");
+                    // filter by identifier of the user instead
                     tokenCacheItems =
-                        tokenCacheItems.Where(
-                                item =>
-                                    !string.IsNullOrEmpty(item.UserAssertionHash) &&
-                                    item.UserAssertionHash.Equals(requestParams.UserAssertion.AssertionHash, StringComparison.OrdinalIgnoreCase))
+                        tokenCacheItems
+                            .Where(item => item.HomeAccountId.Equals(requestParams.Account?.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase))
                             .ToList();
                 }
-                else
-                {
-                    if (!requestParams.IsClientCredentialRequest)
-                    {
-                        requestParams.RequestContext.Logger.Info("Filtering by user identifier...");
-                        // filter by identifier of the user instead
-                        tokenCacheItems =
-                            tokenCacheItems
-                                .Where(item => item.HomeAccountId.Equals(requestParams.Account?.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase))
-                                .ToList();
-                    }
+            }
 
-                    tokenCacheItems = FilterToTenantIdSpecifiedByAuthenticationRequest(requestParams, tokenCacheItems);
```

Comment:
why remove this?

Reply:
BUG BUG:  This is a bug in me restoring my PR after Bogdan restructured to master (it wasn't re-baseable).  I'll need to fix this to not lose this change Jenny did.